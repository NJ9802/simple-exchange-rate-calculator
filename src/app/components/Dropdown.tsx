"use client";

import React, { useState } from "react";
import styled from "styled-components";

interface CurrencyDropdownProps {
  label?: string;
  currencies: [string, number][];
  onSelectCurrency: (currency: [string, number]) => void;
}

const DropdownContainer = styled.div`
  position: relative;
  display: inline-block;
`;

const DropdownButtonWrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 0 0 0 0.75rem;
  border: 1px solid var(--gray-1);
  border-radius: 0.25rem;
`;

const DropdownLabel = styled.label`
  font-weight: bold;
  padding: 0 0.75rem 0 0;
  width: 2.5rem;
  text-align: center;
`;

const DropdownButton = styled.button`
  border: none;
  background-color: var(--gray-1);
  color: var(--gray-3);
  cursor: pointer;
  transition: 300ms;

  &:hover {
    background-color: var(--gray-4);
  }
`;

const DropdownContent = styled.div<{ open: boolean }>`
  display: ${({ open }) => (open ? "block" : "none")};
  position: absolute;
  background-color: var(--gray-2);
  max-height: 12.5rem;
  overflow: auto;
  box-shadow: 0 0.5rem 1rem 0 rgba(0, 0, 0, 0.2);
  z-index: 1;
`;

const DropdownItem = styled.div`
  padding: 0.75rem 1rem;
  cursor: pointer;
  transition: 200ms;

  &:hover {
    background-color: var(--gray-1);
  }
`;

const CurrencyDropdown: React.FC<CurrencyDropdownProps> = ({
  label = "Select",
  currencies,
  onSelectCurrency,
}) => {
  const [open, setOpen] = useState(false);

  const handleDropdownClick = () => {
    setOpen(!open);
  };

  if (!currencies) return;

  return (
    <DropdownContainer>
      <DropdownButtonWrapper>
        <DropdownLabel data-testid="currencyLabel">{label}</DropdownLabel>
        <DropdownButton onClick={handleDropdownClick}>
          {open ? (
            <span className="material-symbols-outlined">arrow_drop_down</span>
          ) : (
            <span className="material-symbols-outlined">arrow_right</span>
          )}
        </DropdownButton>
      </DropdownButtonWrapper>
      <DropdownContent open={open}>
        {currencies.map((currency) => (
          <DropdownItem
            key={currency[0]}
            onClick={() => {
              onSelectCurrency(currency);
              setOpen(false);
            }}
          >
            {currency[0]}
          </DropdownItem>
        ))}
      </DropdownContent>
    </DropdownContainer>
  );
};

export default CurrencyDropdown;
