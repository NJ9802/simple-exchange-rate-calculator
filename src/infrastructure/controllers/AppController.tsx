"use client";

import React, { useState } from "react";
import styled from "styled-components";

import CurrencyDTO from "@/DTOs/currencyDTO";
import CurrencyMapper from "@/DTOs/currencyMapper";
import Button from "@/app/components/Button";
import Dropdown from "@/app/components/Dropdown";
import Header from "@/app/components/Header";
import { Calculator } from "@/core/useCases/exchangeRateCalculator";

import conversionPresenter from "../presenters/ConversionPresenter";

interface AppControllerProps {
  currencies: [string, number][];
}

const AppController: React.FC<AppControllerProps> = ({ currencies }) => {
  const [sourceCurrency, setSourceCurrency] = useState<CurrencyDTO | null>();
  const [targetCurrency, setTargetCurrency] = useState<CurrencyDTO | null>();
  const [amount, setAmount] = useState(0);
  const [conversionResult, setConversionResult] = useState<
    { result: number; exchangeRate: number } | undefined
  >();

  const handleClick = () => {
    if (sourceCurrency && targetCurrency && amount) {
      const exchangeRateCalculator = new Calculator();
      const result = exchangeRateCalculator.convert(
        CurrencyMapper.mapToDomainEntity(sourceCurrency),
        CurrencyMapper.mapToDomainEntity(targetCurrency),
        amount
      );
      setConversionResult(conversionPresenter(result));
    }
  };

  const createCurrencyDTO = (currency: [string, number]) =>
    new CurrencyDTO(currency[0], currency[1]);

  const handleSelectSourceCurrency = (currency: [string, number]) => {
    setSourceCurrency(createCurrencyDTO(currency));
  };

  const handleSelectTargetCurrency = (currency: [string, number]) => {
    setTargetCurrency(createCurrencyDTO(currency));
  };

  return (
    <>
      <Header />
      <CalculatorContainer>
        <ConversionRow>
          <CurrencyContainer>
            <label>FROM</label>

            <Dropdown
              label={sourceCurrency ? sourceCurrency.id : undefined}
              currencies={currencies}
              onSelectCurrency={handleSelectSourceCurrency}
            />
            <Input
              placeholder="Amount"
              type="number"
              onChange={(e) => setAmount(Number(e.target.value))}
            />
          </CurrencyContainer>

          <ReverseButton
            disabled={!sourceCurrency || !targetCurrency}
            onClick={() => {
              setSourceCurrency(targetCurrency);
              setTargetCurrency(sourceCurrency);
              setConversionResult(undefined);
            }}
          >
            <span className="material-symbols-outlined">sync_alt</span>
          </ReverseButton>

          <CurrencyContainer>
            <label>TO</label>

            <Dropdown
              label={targetCurrency ? targetCurrency.id : undefined}
              currencies={currencies}
              onSelectCurrency={handleSelectTargetCurrency}
            />
            <ResultLabel data-testid="resultAmountLabel">
              {conversionResult?.result || "-"}
            </ResultLabel>
          </CurrencyContainer>
        </ConversionRow>

        <ExchangeLabel data-testid="exchangeRateLabel">
          Exchange rate: {conversionResult?.exchangeRate || "-"}
        </ExchangeLabel>

        <ConvertButton
          disabled={!sourceCurrency || !targetCurrency || !amount}
          onClick={handleClick}
        >
          CALCULATE
        </ConvertButton>
      </CalculatorContainer>
    </>
  );
};

const CalculatorContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  max-width: 21.8rem;
  box-shadow: 0 0.5rem 1rem 0 rgba(0, 0, 0, 0.2);
  padding: 1.25rem 1rem;
  margin: 5rem auto;
  border-radius: 0.5rem;
`;

const ConversionRow = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  width: 100%;
`;

const CurrencyContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 1.25rem;
  width: fit-content;
`;

const Input = styled.input`
  width: 3.75rem;
  border: 1px solid #aadae4;
  border-radius: 0.25rem;
  padding: 0.75rem;
  transition: 180ms box-shadow ease-in-out, border-color 180ms ease-in-out;

  &:focus {
    outline: none;
    border-color: var(--blue);
    box-shadow: 0 0 0 3px #2783fb1c;
  }
`;

const ReverseButton = styled(Button)`
  width: 2.5rem;
  height: 2.5rem;
  border-radius: 100%;
`;

const ResultLabel = styled.label`
  display: flex;
  align-items: center;
  height: 37px;
  font-weight: 600;
`;

const ExchangeLabel = styled(ResultLabel)`
  margin-top: 2rem;
`;

const ConvertButton = styled(Button)`
  border-radius: 0.4rem;
  margin-top: 3rem;
  width: 100%;
  padding: 0.8rem 0;
  font-weight: 800;
`;
export default AppController;
