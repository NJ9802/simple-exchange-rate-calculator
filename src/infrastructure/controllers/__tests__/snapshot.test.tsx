import React from "react";
//@ts-ignore
import renderer from "react-test-renderer";
import { currencies } from "@/app/components/__tests__/Dropdown.test";

import AppController from "../AppController";

test("AppController renders correctly", () => {
  const tree = renderer
    .create(<AppController currencies={currencies} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
