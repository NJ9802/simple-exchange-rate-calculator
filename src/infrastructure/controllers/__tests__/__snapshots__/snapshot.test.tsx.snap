// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController renders correctly 1`] = `
[
  <div
    className="sc-iBdoyZ elXume"
  >
    <h1
      className="sc-fsYfxw fasZgm"
    >
      Exchange Rate Calculator
    </h1>
  </div>,
  <div
    className="sc-qZruQ dcTsFz"
  >
    <div
      className="sc-jsEeA-d gMYvCH"
    >
      <div
        className="sc-kFCrIq iLNrCN"
      >
        <label>
          FROM
        </label>
        <div
          className="sc-beyTiQ blxIGe"
        >
          <div
            className="sc-guDMob cVsIDT"
          >
            <label
              className="sc-dmyDmy eqZkmd"
              data-testid="currencyLabel"
            >
              Select
            </label>
            <button
              className="sc-hLQTFJ eKqMAf"
              onClick={[Function]}
            >
              <span
                className="material-symbols-outlined"
              >
                arrow_right
              </span>
            </button>
          </div>
          <div
            className="sc-eDLKEg hSrwAa"
            open={false}
          >
            <div
              className="sc-jTQDJr beNLUK"
              onClick={[Function]}
            >
              USD
            </div>
            <div
              className="sc-jTQDJr beNLUK"
              onClick={[Function]}
            >
              EUR
            </div>
          </div>
        </div>
        <input
          className="sc-irLvoH BcqRO"
          onChange={[Function]}
          placeholder="Amount"
          type="number"
        />
      </div>
      <button
        className="sc-gLLuHO sc-csKJRI hpjupR ORgWz"
        disabled={true}
        onClick={[Function]}
      >
        <span
          className="material-symbols-outlined"
        >
          sync_alt
        </span>
      </button>
      <div
        className="sc-kFCrIq iLNrCN"
      >
        <label>
          TO
        </label>
        <div
          className="sc-beyTiQ blxIGe"
        >
          <div
            className="sc-guDMob cVsIDT"
          >
            <label
              className="sc-dmyDmy eqZkmd"
              data-testid="currencyLabel"
            >
              Select
            </label>
            <button
              className="sc-hLQTFJ eKqMAf"
              onClick={[Function]}
            >
              <span
                className="material-symbols-outlined"
              >
                arrow_right
              </span>
            </button>
          </div>
          <div
            className="sc-eDLKEg hSrwAa"
            open={false}
          >
            <div
              className="sc-jTQDJr beNLUK"
              onClick={[Function]}
            >
              USD
            </div>
            <div
              className="sc-jTQDJr beNLUK"
              onClick={[Function]}
            >
              EUR
            </div>
          </div>
        </div>
        <label
          className="sc-eTNxZ jjRhPM"
          data-testid="resultAmountLabel"
        >
          -
        </label>
      </div>
    </div>
    <label
      className="sc-eTNxZ sc-brSaGm jjRhPM VIRfR"
      data-testid="exchangeRateLabel"
    >
      Exchange rate: 
      -
    </label>
    <button
      className="sc-gLLuHO sc-ktwNLz hpjupR bnoeuw"
      disabled={true}
      onClick={[Function]}
    >
      CALCULATE
    </button>
  </div>,
]
`;
